torch/nn/modules/linear.py

```diff
@@ -73,23 +76,70 @@ def __init__(self, in_features, out_features, bias=True):
         super(Linear, self).__init__()
         self.in_features = in_features
         self.out_features = out_features
-        self.weight = Parameter(torch.Tensor(out_features, in_features))
+
+        sizes = []
+        sizes.append(out_features) \
+            if isinstance(out_features, int) else sizes.extend(out_features)
+        sizes.append(in_features) \
+            if isinstance(in_features, int) else sizes.extend(in_features)
+
+        self.in_prod = self._sizes_prod(in_features)
+        self.out_prod = self._sizes_prod(out_features)
+
+        self.weight = Parameter(torch.empty(sizes))
         if bias:
-            self.bias = Parameter(torch.Tensor(out_features))
+            self.bias = Parameter(torch.empty(self.out_features))
         else:
             self.register_parameter('bias', None)
         self.reset_parameters()
 
+    def _sizes_prod(self, features):
+        assert isinstance(features, (int, Iterable)), (
+            "nn.Linear in_features and out_features must be either an int or "
+            "an Iterable, but got {}."
+        ).format(type(features))
+
+        return features if isinstance(features, int) else \
+            reduce(operator.mul, features, 1) if features else 1
+
     def reset_parameters(self):
-        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
-        if self.bias is not None:
-            fan_in, _ = init._calculate_fan_in_and_fan_out(self.weight)
+        weight, bias = self._reshape_params()
+        init.kaiming_uniform_(weight, a=math.sqrt(5))
+        if bias is not None:
+            fan_in, _ = init._calculate_fan_in_and_fan_out(weight)
             bound = 1 / math.sqrt(fan_in)
-            init.uniform_(self.bias, -bound, bound)
+            init.uniform_(bias, -bound, bound)
+
+    def _reshape_params(self):
+        if self.in_prod != self.in_features or self.out_prod != self.out_features:
+            weight_view = self.weight.view(self.out_prod, self.in_prod)
+            bias_view = self.bias.view(self.out_prod)
+            return weight_view, bias_view
+        else:
+            return self.weight, self.bias
 
     @weak_script_method
     def forward(self, input):
-        return F.linear(input, self.weight, self.bias)
```

Comment:
Is there a reason the logic was done here, as opposed to adjusting the internal implementation of linear to be generalized in this way? Because you did it this way, you had to also duplicate the logic in the C++ module implementation. If you had put it in linear, the C++ module would have gotten the logic as well.

Reply:
I might miss something, it seems to me the existing code has two implementations already: [python](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1390-L1412) and [c++](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Linear.cpp#L14-L28). I can submit another PR to merge these two first, if that is the preferred solution. 

Regarding why not doing it in the `linear` function, I think we can but will need to add some more args to it to specify how many in and out feature dimensions.  Thoughts?

Reply:
I agree with @ezyang's point here.

Reply:
@bddppq can you weigh in here?  I thought we were going to unify these, or do I have that wrong?

Reply:
Thanks for notification. Yes we are trying to directly call the c++ aten::native::linear in the python side. For everyone who is not familiar with the background:
Currently we have two separated but same implementations of linear at [python side F.linear](https://github.com/pytorch/pytorch/blob/7b733e4fc12145583ef96a699cfef5e6b30eee54/torch/nn/functional.py#L1390-L1412) and [c++ side aten::native::linear](https://github.com/pytorch/pytorch/blob/7b733e4fc12145583ef96a699cfef5e6b30eee54/aten/src/ATen/native/Linear.cpp#L14-L28). python module nn.Linear module calls python `F.linear`, so technically c++ linear is not called by any place. Current setup causes some troubles for other backend integration (e.g. mkldnn and tvm, see explanation here https://github.com/pytorch/pytorch/pull/19210#issuecomment-486330064) and we are working on changing python `F.linear` to directly call c++ `aten::native::linear`. @wanchaol has couple onging PRs to make this change.
I would suggest wait for @wanchaol's PRs to land, and then change c++ linear to support the new semantics. If you do not want to wait then you would need to implement the new semantics in both python F.linear and c++ aten linear. In both cases, the nn.Linear's forward function should be kept unchanged.